### PR TITLE
Add a model.check_trainable_weights_consistency

### LINKED ISSUE #8121: [BUG] Number of trainable weights seem to change after model compilation
Consider the following scenario

```python
from keras.layers import Input, Dense
from keras.models import Model
from keras.optimizers import Adam

x = Input(shape=(100,))
y1 = Dense(units=32)(x)
model1 = Model(inputs=x, outputs=y1)
print("MODEL 1")
model1.summary()

model1.trainable = False
x = Input(shape=(100,))
y1 = model1(x)
y2 = Dense(units=64)(y1)
model2 = Model(inputs=x, outputs=y2)
model2.compile(optimizer=Adam(), loss='categorical_crossentropy')

print("MODEL 2")
model2.summary()

model1.trainable = True
print("MODEL 2 after")
model2.summary()
```

I would expect that, since `model2` has been compiled, the output of the two `model2.summary()` calls would be the same. However, after running the above code the actual output is

```
_________________________________________________________________
MODEL 2
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_2 (InputLayer)         (None, 100)               0         
_________________________________________________________________
model_1 (Model)              (None, 32)                3232      
_________________________________________________________________
dense_2 (Dense)              (None, 64)                2112      
=================================================================
Total params: 5,344
Trainable params: 2,112
Non-trainable params: 3,232
_________________________________________________________________
MODEL 2 after
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_2 (InputLayer)         (None, 100)               0         
_________________________________________________________________
model_1 (Model)              (None, 32)                3232      
_________________________________________________________________
dense_2 (Dense)              (None, 64)                2112      
=================================================================
Total params: 5,344
Trainable params: 5,344
Non-trainable params: 0
```

Is this the expected behavior? 

In addition, from other experiments I get the feeling that although the number of trainable weights reported by `summary()` differs in the two cases, the actual number of trainable weights is the expected one, that is, in the example above, even after setting `model1.trainable=True`, training will really update only 2,112 parameters and not 5,344. So, maybe this is just a reporting issue.

### PR DESCRIPTION
**Pull request by julienr**
This will raise a UserWarning when the user modifies model.trainable
and tries to print a model summary or launch a fit without having
called .compile.

Calling .compile() is necessary because trainable weights are collected
in compile (model._collected_trainable_weights).

Fixes #8121

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -945,9 +946,25 @@ def handle_metrics(metrics, weights=None):
...
 
+    def check_trainable_weights_consistency(self):
+        """
+        Checks that trainable_weights and _collected_trainable_weights are
+        consistent
+        """
+        if not hasattr(self, '_collected_trainable_weights'):
+            return
+
+        if count_params(self.trainable_weights) != \
```
Comment (fchollet) [inline keras/engine/training.py]: Please no `\`

File: keras/engine/training.py [hunk 2]
```diff
@@ -945,9 +946,25 @@ def handle_metrics(metrics, weights=None):
...
+           count_params(self._collected_trainable_weights):
+            warnings.warn(UserWarning(
+                'Discrepancy between trainable weights and collected trainable'
+                ' weights, did you set model.trainable without calling'
```
Comment (fchollet) [inline keras/engine/training.py]: Nit: ` around code keywords

File: keras/utils/layer_utils.py [hunk 3]
```diff
@@ -5,6 +5,11 @@
 import numpy as np
 
 
+def count_params(weights):
+    """Count the total number of scalars composing the weights"""
```
Comment (fchollet) [inline keras/utils/layer_utils.py]: Docstring formatting: one-line summary should end with a period; the dosctring should include a `# Arguments` and `# Returns` section.

**Comment by julienr** [comment]
Thanks for the quick review. I think I addressed your comments.

File: keras/engine/training.py [hunk 4]
```diff
@@ -945,9 +946,25 @@ def handle_metrics(metrics, weights=None):
         trainable_weights = self.trainable_weights
         self._collected_trainable_weights = trainable_weights
 
+    def check_trainable_weights_consistency(self):
```
Comment (fchollet) [inline keras/engine/training.py]: Please make this method private (underscore) and make the docstring style-compliant.

**Pull request merged**